From: Steffen Trumtrar <s.trumtrar@pengutronix.de>
Date: Tue, 8 Mar 2016 13:46:14 +0100
Subject: [PATCH] evmctl: add parameter -e to set evm hash algo

The paramter -a sets the hash algorithm only for IMA. To not break
anything, add a new parameter -e to be able to change the hash for
EVM, too.

Signed-off-by: Steffen Trumtrar <s.trumtrar@pengutronix.de>
---
 src/evmctl.c    | 27 +++++++++++++++++++++++----
 src/imaevm.h    |  1 +
 src/libimaevm.c |  1 +
 3 files changed, 25 insertions(+), 4 deletions(-)

diff --git a/src/evmctl.c b/src/evmctl.c
index 559e4cbf9176..d7352d87ef71 100644
--- a/src/evmctl.c
+++ b/src/evmctl.c
@@ -319,6 +319,7 @@ static int calc_evm_hash(const char *file, unsigned char *hash)
 	char uuid[16];
 	struct h_misc_64 hmac_misc;
 	int hmac_size;
+	const EVP_MD *md;
 
 	if (lstat(file, &st)) {
 		log_err("Failed to stat: %s\n", file);
@@ -350,7 +351,13 @@ static int calc_evm_hash(const char *file, unsigned char *hash)
 		return -1;
 	}
 
-	err = EVP_DigestInit(&ctx, EVP_sha1());
+	md = EVP_get_digestbyname(params.evm_hash_algo);
+	if (!md) {
+		log_err("EVP_get_digestbyname() failed\n");
+		return 1;
+	}
+
+	err = EVP_DigestInit(&ctx, md);
 	if (!err) {
 		log_err("EVP_DigestInit() failed\n");
 		return 1;
@@ -454,7 +461,7 @@ static int sign_evm(const char *file, const char *key)
 	if (len <= 1)
 		return len;
 
-	len = sign_hash("sha1", hash, len, key, NULL, sig + 1);
+	len = sign_hash(params.evm_hash_algo, hash, len, key, NULL, sig + 1);
 	if (len <= 1)
 		return len;
 
@@ -860,6 +867,7 @@ static int calc_evm_hmac(const char *file, const char *keyfile, unsigned char *h
 	ssize_t list_size;
 	struct h_misc_64 hmac_misc;
 	int hmac_size;
+	const EVP_MD *md;
 
 	key = file2bin(keyfile, NULL, &keylen);
 	if (!key) {
@@ -905,7 +913,13 @@ static int calc_evm_hmac(const char *file, const char *keyfile, unsigned char *h
 		goto out;
 	}
 
-	err = !HMAC_Init(&ctx, evmkey, sizeof(evmkey), EVP_sha1());
+	md = EVP_get_digestbyname(params.evm_hash_algo);
+	if (!md) {
+		log_err("EVP_get_digestbyname() failed\n");
+		return 1;
+	}
+
+	err = !HMAC_Init(&ctx, evmkey, sizeof(evmkey), md);
 	if (err) {
 		log_err("HMAC_Init() failed\n");
 		goto out;
@@ -1464,6 +1478,7 @@ static void usage(void)
 	printf(
 		"\n"
 		"  -a, --hashalgo     sha1 (default), sha224, sha256, sha384, sha512\n"
+		"  -e, --evmhashalgo  sha1 (default), sha224, sha256, sha384, sha512\n"
 		"  -s, --imasig       make IMA signature\n"
 		"  -d, --imahash      make IMA hash\n"
 		"  -f, --sigfile      store IMA signature in .sig file instead of xattr\n"
@@ -1508,6 +1523,7 @@ static struct option opts[] = {
 	{"imasig", 0, 0, 's'},
 	{"imahash", 0, 0, 'd'},
 	{"hashalgo", 1, 0, 'a'},
+	{"evmhashalgo", 1, 0, 'e'},
 	{"pass", 2, 0, 'p'},
 	{"sigfile", 0, 0, 'f'},
 	{"uuid", 2, 0, 'u'},
@@ -1565,7 +1581,7 @@ int main(int argc, char *argv[])
 	g_argc = argc;
 
 	while (1) {
-		c = getopt_long(argc, argv, "hvnsda:p::fu::k:t:ri", opts, &lind);
+		c = getopt_long(argc, argv, "hvnsda:e:p::fu::k:t:ri", opts, &lind);
 		if (c == -1)
 			break;
 
@@ -1591,6 +1607,9 @@ int main(int argc, char *argv[])
 		case 'a':
 			params.hash_algo = optarg;
 			break;
+		case 'e':
+			params.evm_hash_algo = optarg;
+			break;
 		case 'p':
 			if (optarg)
 				params.keypass = optarg;
diff --git a/src/imaevm.h b/src/imaevm.h
index 711596c3f3fa..ef7858b8faa0 100644
--- a/src/imaevm.h
+++ b/src/imaevm.h
@@ -178,6 +178,7 @@ struct libevm_params {
 	int verbose;
 	int x509;
 	const char *hash_algo;
+	const char *evm_hash_algo;
 	const char *keyfile;
 	const char *keypass;
 };
diff --git a/src/libimaevm.c b/src/libimaevm.c
index 1c5da965468c..595908395514 100644
--- a/src/libimaevm.c
+++ b/src/libimaevm.c
@@ -129,6 +129,7 @@ struct libevm_params params = {
 	.verbose = LOG_INFO - 1,
 	.x509 = 1,
 	.hash_algo = "sha1",
+	.evm_hash_algo = "sha1",
 };
 
 static void __attribute__ ((constructor)) libinit(void);
